src/ipa_pc/mod.rs

```diff
@@ -1204,146 +1222,204 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+        let (h_commitments, h_randomnesses) = Self::commit(
+            &ck,
+            vec![&LabeledPolynomial::new(format!("h_poly"), h_polynomial.clone(), None, None)],
+            if has_hiding {
+                if rng.is_none() {
+                    Err(Error::Other("Rng not set".to_owned()))?
+                }
+                Some(rng.as_mut().unwrap())
+            } else {
+                None
             }
-            Randomness::rand(segments_count, false, rng.as_mut().unwrap())
-        } else {
-            Randomness::empty(segments_count)
-        };
+        )?;
 
-        // Commitment of the h(X) polynomial
-        let commit_time = start_timer!(|| format!("Commit to batch polynomial of degree {}", batch_polynomial.degree()));
-        let batch_commitment: Vec<G>;
+        let h_comm = h_commitments[0].commitment().comm.clone();
+        let h_randomness = h_randomnesses[0].randomness();
 
-        if p_len > key_len {
+        end_timer!(commit_time);
 
-            batch_commitment = (0..segments_count).into_iter().map(
-                |i| {
-                    Ok(Self::cm_commit(
-                        &ck.comm_key,
-                        &batch_polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)],
-                        Some(ck.s),
-                        Some(batch_randomness.rand[i]),
-                    )?.into_affine())
+        let open_time = start_timer!(|| "Open LC(p_1(X),p_2(X),...,p_m(X),h(X)) polynomial");
+
+        // Fresh random challenge x for multi-point to single-point reduction.
+        // Except the `batch_commitment`, all other commitments are already bound 
+        // to the internal state of the Fiat-Shamir
+        fs_rng.absorb(&to_bytes![h_comm].unwrap());
+        let x_point: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
+
+        // LC(p_1(X),p_2(X),...,p_m(X),h(X))
+        let mut lc_polynomial = Polynomial::zero();
+        let mut lc_randomness = Randomness::empty(0);
+
+        // LC(C): reconstructed commitment to LC(p_1(X),p_2(X),...,p_m(X),h(X))
+        let mut lc_comm = vec![];
+
+        let mut cur_challenge = G::ScalarField::one();
+
+        for (label, (_point_label, point)) in query_set.iter() {
+            let labeled_polynomial =
+                *poly_map.get(label).ok_or(Error::MissingPolynomial {
+                    label: label.to_string(),
+                })?;
+
+            let labeled_commitment =
+                *commitment_map.get(label).ok_or(Error::MissingCommitment {
+                    label: label.to_string(),
+                })?;
+
+            let labeled_randomness =
+                *rands_map.get(label).ok_or(Error::MissingRandomness {
+                    label: label.to_string(),
+                })?;
+
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                (G::ScalarField::zero() - point),
```

Comment:
-point instead of  (G::ScalarField::zero() - point) ?

Reply:
Done